File: src/simulation/space/WorldMap.java
package simulation.space;

import simulation.Controller;
import simulation.space.tile.Tile;
import simulation.space.tile.TileTagSetterKt;

import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static simulation.Controller.*;

/**
 * Represents tile map of the world
 */
public class WorldMap {
    private List<List<Tile>> tiles;
    private int x;
    private int y;
    private List<TectonicPlate> tectonicPlates = new ArrayList<>();

    public WorldMap(List<List<Tile>> tiles) {
        this.tiles = tiles;
        x = tiles.size();
        y = tiles.get(0).size();
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public List<List<Tile>> getLinedTiles() {
        return tiles;
    }

    public List<TectonicPlate> getTectonicPlates() {
        return tectonicPlates;
    }

    public void addPlate(TectonicPlate plate) {
        tectonicPlates.add(plate);
    }

    public Tile get(int x, int y) {
        if (SpaceData.INSTANCE.getData().getXMapLooping()) {
            x = cutCoordinate(x, getX());
        } else if (!checkCoordinate(x, getX())) {
            return null;
        }
        if (SpaceData.INSTANCE.getData().getYMapLooping()) {
            y = cutCoordinate(y, getY());
        } else if (!checkCoordinate(y, getY())) {
            return null;
        }
        return getLinedTiles().get(x).get(y);
    }

    private int cutCoordinate(int coordinate, int max) {
        if (coordinate < 0) {
            coordinate = coordinate % max + max;
        }
        return coordinate % max;
    }

    private boolean checkCoordinate(int coordinate, int max) {
        return coordinate >= 0 && coordinate < max;
    }

    public void setTags() {
        int name = 0;
        List<Tile> allTiles = getTiles();
        for (Tile tile : allTiles) {
            if (TileTagSetterKt.setTags(tile, "" + name)) {
                name++;
            }
        }
    }

    public List<Tile> getTiles() {
        return tiles.stream()
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
    }

    public List<Tile> getTiles(Predicate<Tile> predicate) {
        return tiles.stream()
                .flatMap(Collection::stream)
                .filter(predicate)
                .collect(Collectors.toList());
    }

    public synchronized void update() {//TODO parallel
        for (List<Tile> line : getLinedTiles()) {
            for (Tile tile : line) {
                tile.startUpdate();
            }
        }

        for (List<Tile> line : getLinedTiles()) {
            for (Tile tile : line) {
                tile.middleUpdate();
            }
        }
    }

    public synchronized void finishUpdate() {
        for (List<Tile> line : getLinedTiles()) {
            for (Tile tile : line) {
                tile.finishUpdate();
            }
        }
    }

    public void geologicUpdate() {
        for (List<Tile> line : getLinedTiles()) {
            for (Tile tile : line) {
                tile.levelUpdate();
            }
        }
        platesUpdate();
    }

    public void platesUpdate() {
        for (TectonicPlate plate : getTectonicPlates()) {
            plate.move();
        }
    }
}

File: EcosystemSimulation/src/shmp/simulation/space/WorldMap.kt
package shmp.simulation.space

import shmp.simulation.space.SpaceData.data
import shmp.simulation.space.tile.Tile
import shmp.simulation.space.tile.setTags


class WorldMap(val linedTiles: List<List<Tile>>) {
    val maxX = linedTiles.size
    val maxY = linedTiles[0].size

    val tectonicPlates = mutableListOf<TectonicPlate>()

    fun addPlate(plate: TectonicPlate) {
        tectonicPlates.add(plate)
    }

    operator fun get(_x: Int, _y: Int): Tile? {
        var curX = _x
        var curY = _y

        if (data.xMapLooping)
            curX = cutCoordinate(curX, maxX)
        else if (!checkCoordinate(curX, maxX))
            return null

        if (data.yMapLooping)
            curY = cutCoordinate(curY, maxY)
        else if (!checkCoordinate(curY, maxY))
            return null

        return linedTiles[curX][curY]
    }

    private fun cutCoordinate(coordinate: Int, max: Int): Int {
        var curCoordinate = coordinate

        if (curCoordinate < 0)
            curCoordinate = curCoordinate % max + max

        return curCoordinate % max
    }

    private fun checkCoordinate(coordinate: Int, max: Int) = coordinate in 0 until max

    fun setTags() {
        var name = 0
        val allTiles = tiles

        for (tile in allTiles)
            if (setTags(tile, "" + name))
                name++
    }

    val tiles = linedTiles.flatten()

    fun getTiles(predicate: (Tile) -> Boolean) = linedTiles
            .flatten()
            .filter(predicate)

    @Synchronized
    fun update() { //TODO parallel
        for (line in linedTiles)
            for (tile in line)
                tile.startUpdate()

        for (line in linedTiles)
            for (tile in line)
                tile.middleUpdate(this)
    }

    @Synchronized
    fun finishUpdate() {
        for (line in linedTiles)
            for (tile in line)
                tile.finishUpdate()
    }

    fun geologicUpdate() {
        for (line in linedTiles)
            for (tile in line)
                tile.levelUpdate()

        platesUpdate()
    }

    fun platesUpdate() {
        for (plate in tectonicPlates)
            plate.move()
    }
}
